### DISCUSSION TITLE
Guidance for integrating an Audio System

### DISCUSSION THREAD

**Discussion by na2axl** [ezEngine/ezEngine#610] [Q&A] [answered]
Hey there,

I'm totally new in ezEngine, but after used it for few days now I think it's a great piece and I want to give it a real try. My first plan is to integrate my audio engine [here](https://github.com/SparkyStudios/AmplitudeAudioSDK) into ezEngine and replace the default one based on FMOD. I've seen the engine is plugin based and in the documentation there is a little explanation on how to create plugins, but how is the engine is expecting an audio backend to be integrated? Should I implement/extends a specific interface/class or just create custom components that use my engine?

Any help/tip appreciated 😄

**Comment by jankrassnigg** [comment, marked as answer]
Hey, really interesting project!

It is best to follow along what the fmod integration is doing. That should give you pretty good guidance.

There are 2 plugins for fmod:
* EnginePlugins\FmodPlugin
* EditorPlugins\Fmod\EditorPluginFmod

The first one is the more important one, as it contains all of the integration into the game engine. The second one is for adding assets to the editor, which are used to set configure sound banks and expose the fmod sound events in the asset browser. To get started you can skip all that and just hardcode some stuff.

The only thing that's important from the EditorPluginFmod is that in its CMakeLists.txt it copies the "Fmod.ezPluginBundle" file to the binary directory.  That file is important so that fmod shows up as a [selectible plugin](https://ezengine.net/pages/docs/projects/plugin-selection.html). You would want to do the same on your plugin, so that you can activate your sound engine (and deactivate fmod).

Let's look at FmodPlugin.

In FmodSingleton.h the class "ezFmod" implements ezSoundInterface. You should do the same. This interface is VERY basic, you can't do much through it. The most interesting functions for you are ezSoundInterface::UpdateSound and ezSoundInterface::SetListener. The latter is used by the editor to override the listener position with the editor camera position.

The rest is mostly done through components and resources.

ezFmodListenerComponent represents your listener. ezFmodEventComponent represents one sound instance. It's probably easy to copy and paste those two components and use them to talk to your sound engine.

That's really everything you need to get started. The editor will show the component properties that are listed in the reflection macro blocks (see FmodEventComponent.cpp lines 230+). That's how you can make stuff configurable to start.

Once you want to make things nicer for the user, you would need to have a resource type for your sound bank (see ezFmodSoundBankResource). This is where the fmod integration becomes a bit more complicated, because in EZ we want to represent every sound effect as a separate resource (ezFmodSoundEventResource), even though all sounds are packed into single big sound banks. Therefore the two resource types exist, but the event resource mostly just points into the sound bank resource.

Finally, if you want editor integration for sound banks and events, you would need to create an "asset type" (see ezSoundBankAssetDocument and ezSoundEventAssetDocument) and lots of scaffolding around that to connect everyting up properly. I would advise to concentrate on the rest first, though. This can be done, once you have a working prototype.

Hope that helps!

**Reply by na2axl** [reply]
Thanks for the very good explanation. I'll follow that path and let you know of the progress 👍🏾

**Reply by na2axl** [reply]
Hey @jankrassnigg, thanks to your tip I finally succeed to integrate my audio engine 😃. But It's still a basic implementation, for now I'm just triggering audio and gathering occlusion/obstruction data from ray casting.

I had to reproduce a lot of code from the FMOD plugin, which I think can be offered by the engine itself as a _base_ for every audio implementation. For that I was wondering if you are interested in having full ATL (Audio Translation Layer) support right in ezEngine. I've started to write a plugin (nothing fancy yet 😅) to _abstract_ every aspect of an audio middleware, so it will be easy to anyone to integrate other audio engines (like Wwise) quickly to the game engine without writing a lot of code.

If needed, I can write an RFC for that.

**Reply by jankrassnigg** [reply]
Congratulations on your first success! That was pretty fast, actually.

I'm wondering how much you can abstract away and how much duplicate work can be saved. Of course you have to duplicate the component stuff, but I thought they would go pretty directly to fmod with little in between. So I'm intruiged that you say there is room for an abstraction layer.

"RFC" sounds so formal and like a waste of time :D I would rather be interested to just see actual code. So if you have a basic outline for this ATL, just send me a link to a GitHub branch or so.

However, I think it shouldn't be a mandatory layer that every sound engine has to build upon. Rather it could be a static library, that a sound plugin can link against, so that it already has base functionality available. But if someone doesn't want that, it's entirely optional.

Anyway, glad to hear that you are having a good time with EZ so far :-)

**Reply by na2axl** [reply]
Hey @jankrassnigg, I'm currently cooking the thing, once having some near good shape I'll not miss to share it with you :-)

But so far, I'm definitely not planning to replace the actual `ezSoundInterface` by ATL, instead, the ATL will extend that interface, and audio middleware will extend the ATL (if they want to).

I'm taking advantage of the plugin system, the ATL itself will be a plugin, so it will not be a mandatory layer but something that the user can enable or disable as their wishes. Also, I've seen that a plugin can inform the engine that it depends on another plugin, I've not checked how it works yet but I think when an audio middleware will want to use the ATL it will just add the ATL plugin as a dependency.

The only thing I've changed as of now in the `Core` library is the `ezSurfaceResourceDescriptor`, to add the amount of `sound obstruction` a surface provides when sound rays pass through it. This allowed me to have proper results in the sound obstruction/occlusion calculation according to what the player sees in the game (wired wall surface vs stone wall surface).

Hopefully I will share some code to you next week! 🤞🏾

**Reply by jankrassnigg** [reply]
Looking forward to it!

Extending the surfaces is exactly the right thing for this kind of interaction, btw.

**Reply by na2axl** [reply]
Hi @jankrassnigg, after a lot of tests and different implementations, I think now I following a very good path on that implementation! It took me a while to understand how EZ works internally, with assets and all, but still, there is how the AudioSystem plugin looks now https://github.com/ezEngine/ezEngine/compare/dev...SparkyStudios:ezEngine:plugin/audio-system

The actual implementation lacks of several features but it is working, I'm able to load the audio system, register a middleware, load audio controls, and trigger an audio event. For now I'm doing tests using Amplitude as the audio middleware, after the AudioSystem will be more mature, I will try to implement a middleware with FMOD or Wwise too.

I should probably write some description of what is going on there, I will do it sometime, but in the while, I'm here if any questions!

I'm not sure on how to track our discussion about it... Tell me if it's ok to open a draft pull request :-)

**Reply by jankrassnigg** [reply]
Awesome 👍I'll take a look as soon as I can, but not sure when exactly. You can open a draft pull request, if you like, no problem at all.

**Reply by jankrassnigg** [reply]
Wow, that's quite a lot of code. I've now gone through it and I think I have a much clearer picture now.

First of all, let me say that I'm quite impressed with the quality of your work. You've managed to use the EZ infrastructure pretty much exactly how it is supposed to be used. For someone with no prior knowledge of EZ, that's quite amazing. You obsiously also know way more about audio programming than I do, and I would really love to see Amplitude well integrated into EZ.

Looking through the PR, I think I see the following design goals:

1. Have audio features exposed through a fixed set of components and resources, so that users interact with audio in EZ always the same way, no matter which middleware is used.
2. Have a common code interface, through which audio middleware can be plugged in.

Number 2 is a necessatiy to achieve 1.

Neither 1) nor 2) are currently how EZ is doing it for audio. We are also not doing it that way for physics for which we actually have two implementations (PhysX + Jolt). If you were to switch a project from PhysX to Jolt or vice versa, you would need to manually adjust every prefab to use different components, which is quite tedius.

Putting an abstraction layer in between is the logical way to "solve" this issue.

However, not having much of an abstraction layer actually is a deliberate decision. What you are proposing is, for my taste, going too far, and I want to explain why I think so.

In general 1) is a nice goal. I have worked on different engines and especially when you give out an engine to "customers" whose use cases you don't know, it is always great when people select whatever they want and "it just works". In my experience, though, it is neither what people actually need, nor does it ever work as well as intended.

Both audio and physics engines are very deceptive here. No matter which one you look at, their feature set looks pretty much identical, so an abstraction should be easy. But once you actually do it, you notice that they often differ quite dramatically in how you would achieve different things. For example in FMOD environmental effects are all done through their very generic sound event + parameter interface, whereas other engines may do those things through global configurations. That means what effects are possible and how you achieve them is fundamentally different on a technical level.

In practice this has several consequences:

a) The abstraction layer will be designed towards the API(s) that you start with. A good example is our graphcis abstraction layer. It's close to a 1:1 copy of D3D11, because that was for a long time what we worked with exclusively. Now that we are porting to Vulkan, we have to refactor quite a lot of that. This is ok here, because we will put in the work, but as an abstraction for the future it actually failed, because other people were not able to just implement it with Vulkan and have it work. So in practice it is more of an abstraction-after-the-fact, where you can talk to all of the existing implementations, but reaching the goal to make it future proof is extraordinary hard (and impossible to know before hand).

b) The abstraction layer will lean towards providing the lowest common denominator both in its feature set and usability. As an example in PhysX it is very easy to set different materials for each shape in a rigid body. In Jolt this is not supported. So what would an abstraction do? Either you go Jolt's way and limit what is possible in PhysX, or you expose what PhysX does, but then when using Jolt the documentation just says that "only the first shape's material will be used for all shapes" ? Either way, the user experience isn't what it should be.

c) If the abstraction is too different from the underlying API, education becomes a problem as well. FMOD, for instance, has really good tutorials how to achieve various effects. And you can do all sorts of complicated things through their rather elegant, minimal API. Most of the things are done in FMOD Studio, and in code you just start a few events and set parameters. With an abstraction in between that works differently, people can't easily apply this knowledge and instead need to consult additional documentation to figure out how to map the knowledge from the tutorial to the engine. To me, this is a big concern, because neither do I have the time to support people much, nor do I even have that kind of expertise myself, because I'm no FMOD expert either.

d) For good user experience you need good tools. If people should have the great FMOD user experience with all backends, you need to implement many of their tools yourself, as well. This is theoretically possible when you have a big team, but even then, you often just duplicate everything in lower quality and don't benefit from either FMOD's or WWise's tools (etc).

Looking at the audio abstraction, I see several things for which I do not see an equivalent in FMOD. On the other hand, you are adding an asset for mapping things from the abstraction to the implementation (I guess). So if I understand this correctly, if I add a sound event to a bank in FMOD, I would now also add this to the asset in EZ before I can use it? That would be inferior to what we currently have, where all you need to do is transform the sound bank, and an asset for every event inside it automatically shows up and is usable right away. Here users need to do more work, with no real benefit.

For programmers abstractions are second nature. We see things that are similar, and want to make them the same. The thing is, it is extremely rare that people actually want to switch between systems and it isn't done on a whim. The "because we can" argument means very little to people who actually make games. On the other hand, if an abstraction limits your available feature set, or adds friction in usability, that affects everyone.

Preventing code duplication is a very programmer way of thinking as well. Users don't care that you copy&pasted 5000 lines of code. On the other hand, they do care, if even just one feature specific to one API is not available, because it isn't what all the other APIs provide. This difference can be the killer argument why they would want to use that thing and not the others. And if you look at other abstractions, they often bend over backwards to somehow expose the implementation specific functionality where it's really needed (e.g. Unreal abstracts physics and has a lot of PhysX / Chaos specific loopholes in there).
Abstractions make sense for example for platform ports. My mouse and keyboard work the same on Windows, Linux etc, and I don't want to use different code to access them. So here an abstraction can actually guarantee me, that I get the same behavior and features on all platforms, so it is a clear benefit with no compromises.

These are all lessons I had to learn the hard way over many years, and I've given you this very elaborate answer because for all the effort that you've put into this you deserve a detailed explanation why I'm not happy with this approach.

That is also why I originally wrote "I'm wondering how much you can abstract away and how much duplicate work can be saved." because I didn't expect you to go all-in with a complete abstraction on all levels.


Now, where do we go from here?

As I already said, I would really love to have your audio engine integrated into EZ. However, I wouldn't continue the road to try to build an abstraction around audio in general. Having some duplicated code is fine. You'll notice at some point, that having the code duplicated, allows you to adjust many things so that they fit perfectly to your API, where with an abstraction to adhere to, you either couldn't do something at all, or would have to worry about how every other API could handle those things as well. So this gives you much more freedom to make your audio integration the best way possible (and I can also help with that, for instance on the editor usability side).

Now there are a few things where one could add abstractions that help other audio integrations. For example you mentioned computations for obstruction and occlusion. This is actually a very interesting topic, where I am looking forward to your expertise. Generally, as I understand it, you mostly need those two float values but there are many ways to compute them. You could use physics raycasts, or a separate representation, such as [Steam Audio](https://steamcommunity.com/app/596420). I originally thought that your work might go into such a direction, because here you can certainly share a lot of code. E.g. you could have a base interface to register sources that should get occlusion, and then different implementations that provide you with the result. All audio engines would then benefit from this.

I hope you are not angry with this feedback and see it all as a waste of time, but rather as part of the process to get to a good solution for everyone. It always pains me when I need to tell people that all their hard work is not going to be accepted in this form. With people from our team I often discuss for hours about how to design something, with outside collaborators this isn't really possible, so it's more a process of "here is a code proposal" - "here is my feedback on that".

Long story short, I hope what I've written above makes sense to you.

**Reply by na2axl** [reply]
Hey @jankrassnigg. With all the edited files growing I could feel it 😅. But no worries, I should have consulted you before digging into this, now I made you write all this text, my bad.... 😓 

As mentioned in my PR, I've started that implementation following my experience on CryEngine and O3DE, which from my thoughts have really good audio abstraction that fits with most of APIs. My goal was not to replace how audio works in EZ or oblige the end user to depend on the ATL, but instead, provide something "which does the job" to users who doesn't want to implement components or create any resources to integrate an audio engine. With the good plugin management that EZ is proposing I was thinking it's something achievable without breaking what EZ is now, in a way that if a user doesn't want it, he just disable that plugin and go straight with a custom implementation like what FMOD is now. If he feel comfortable with, he just enable the plugin and implement the middleware interface...

But I clearly understand your point, and you are right with the possible issues when trying to abstract different software APIs. But I can say, this is the cost for such a thing 😅 

I am currently working on a project with EZ (I think I will share some screenshots in #show-and-tell sometime), that's why the time spent on that, so no worries, I can't take it as a waste of time 🙂. I plan to stick with this implementation for personal usage since it have some benefits I want (audio things are done on a separated thread for example), but yes I will be happy to share my experience with you on that occlusion/obstruction computation 👍🏾 

Anyway, I've learned a lot on EZ with this, and I feel really comfortable with the engine, you really did a good job here 🚀
